[BUG] assignment of read-only member -- occurs with cython tuple optimizations

### Describe the bug

I was helping pyav update to cython 3 https://github.com/PyAV-Org/PyAV/pull/1145

and I hit the following

### Code to reproduce the behaviour:

```
# activate your dev environment
mamba install cython ffmpeg=5 numpy pytest
git clone git@github.com:hmaarrfk/PyAV.git
cd PyAV
git checkout cython3
pip install -e . -vv --no-build-isolation --no-deps
```

Apply the patch:
```patch
diff --git a/av/video/format.pyx b/av/video/format.pyx
index 9ad5391..b966582 100644
--- a/av/video/format.pyx
+++ b/av/video/format.pyx
@@ -47,14 +47,10 @@ cdef class VideoFormat(object):
         self.ptr = lib.av_pix_fmt_desc_get(pix_fmt)
         self.width = width
         self.height = height
-        # hmaarrfk -- 2023/07/23
-        # Note on tuple([])
-        # Cython 3 seems to have trouble with cdef tuples, so we use a list
-        # it complains about some const identifier not being able to get assigned
-        self.components = tuple([
+        self.components = tuple(
             VideoFormatComponent(self, i)
             for i in range(self.ptr.nb_components)
-        ])
+        )
 
     def __repr__(self):
         if self.width or self.height:
```


```
  /home/mark/mambaforge/envs/dev/bin/x86_64-conda-linux-gnu-cc -Wno-unused-result -Wsign-compare -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /home/mark/mambaforge/envs/dev/include -fPIC -O2 -isystem /home/mark/mambaforge/envs/dev/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/mark/mambaforge/envs/dev/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/mark/mambaforge/envs/dev/include -fPIC -I/home/mark/mambaforge/envs/dev/include -I/home/mark/mambaforge/envs/dev/include/python3.9 -c src/av/video/format.c -o /tmp/tmpmwp8qucp.build-temp/src/av/video/format.o
  src/av/video/format.c: In function '__pyx_pf_2av_5video_6format_11VideoFormat_5_init_genexpr':
  src/av/video/format.c:3310:40: error: assignment of read-only member '__pyx_genexpr_arg_0'
   3310 |   __pyx_cur_scope->__pyx_genexpr_arg_0 = __pyx_genexpr_arg_0;
        |                          
```

Log and format.c.txt from 3.0.0 but I confirmed that it also occurs in main branch
* [log.txt](https://github.com/cython/cython/files/12139884/log.txt)
* [format.c.txt](https://github.com/cython/cython/files/12139894/format.c.txt)


### Expected behaviour

that it compiles. I worked around it by using a list comprehension before the `tuple` call.

### OS

Linux

### Python version

3.9.16

### Cython version

main or 3.0.0 (conda-forge)

### Additional context

_No response_

--- a/Cython/Compiler/ExprNodes.py
+++ b/Cython/Compiler/ExprNodes.py
@@ -3294,7 +3294,7 @@ class NextNode(AtomicExprNode):
             return iterator_type.base_type
         elif iterator_type.is_cpp_class:
             item_type = env.lookup_operator_for_types(self.pos, "*", [iterator_type]).type.return_type
-            item_type = PyrexTypes.remove_cv_ref(item_type, dont_remove_fakeref=False)
+            item_type = PyrexTypes.remove_cv_ref(item_type, remove_fakeref=True)
             return item_type
         else:
             # Avoid duplication of complicated logic.

--- a/Cython/Compiler/ExprNodes.py
+++ b/Cython/Compiler/ExprNodes.py
@@ -3294,10 +3294,7 @@ class NextNode(AtomicExprNode):
             return iterator_type.base_type
         elif iterator_type.is_cpp_class:
             item_type = env.lookup_operator_for_types(self.pos, "*", [iterator_type]).type.return_type
-            if item_type.is_reference:
-                item_type = item_type.ref_base_type
-            if item_type.is_cv_qualified:
-                item_type = item_type.cv_base_type
+            item_type = PyrexTypes.remove_cv_ref(item_type, dont_remove_fakeref=False)
             return item_type
         else:
             # Avoid duplication of complicated logic.

--- a/Cython/Compiler/ParseTreeTransforms.py
+++ b/Cython/Compiler/ParseTreeTransforms.py
@@ -1720,14 +1720,20 @@ class _HandleGeneratorArguments(VisitorTransform, SkipDeclarations):
                 cname = EncodedString(Naming.genexpr_arg_prefix + Symtab.punycodify_name(str(name_source)))
                 name_decl = Nodes.CNameDeclaratorNode(pos=pos, name=name)
                 type = node.type
-                if type.is_reference and not type.is_fake_reference:
-                    # It isn't obvious whether the right thing to do would be to capture by reference or by
-                    # value (C++ itself doesn't know either for lambda functions and forces a choice).
-                    # However, capture by reference involves converting to FakeReference which would require
-                    # re-analysing AttributeNodes. Therefore I've picked capture-by-value out of convenience
-                    # TODO - could probably be optimized by making the arg a reference but the closure not
-                    # (see https://github.com/cython/cython/issues/2468)
-                    type = type.ref_base_type
+                while ((type.is_reference and not type.is_fake_reference) or
+                        type.is_cv_qualified):
+                    if type.is_cv_qualified:
+                        # strip away cv types - they shouldn't be applied to the
+                        # function argument or to the closure struct.
+                        type = type.cv_base_type
+                    else:
+                        # It isn't obvious whether the right thing to do would be to capture by reference or by
+                        # value (C++ itself doesn't know either for lambda functions and forces a choice).
+                        # However, capture by reference involves converting to FakeReference which would require
+                        # re-analysing AttributeNodes. Therefore I've picked capture-by-value out of convenience
+                        # TODO - could probably be optimized by making the arg a reference but the closure not
+                        # (see https://github.com/cython/cython/issues/2468)
+                        type = type.ref_base_type
 
                 name_decl.type = type
                 new_arg = Nodes.CArgDeclNode(pos=pos, declarator=name_decl,

--- a/Cython/Compiler/ParseTreeTransforms.py
+++ b/Cython/Compiler/ParseTreeTransforms.py
@@ -1729,7 +1729,7 @@ class _HandleGeneratorArguments(VisitorTransform, SkipDeclarations):
                 # re-analysing AttributeNodes. Therefore I've picked capture-by-value out of convenience
                 # TODO - could probably be optimized by making the arg a reference but the closure not
                 # (see https://github.com/cython/cython/issues/2468)
-                type = PyrexTypes.remove_cv_ref(type, dont_remove_fakeref=True)
+                type = PyrexTypes.remove_cv_ref(type, remove_fakeref=False)
 
                 name_decl.type = type
                 new_arg = Nodes.CArgDeclNode(pos=pos, declarator=name_decl,

--- a/Cython/Compiler/ParseTreeTransforms.py
+++ b/Cython/Compiler/ParseTreeTransforms.py
@@ -1720,20 +1720,16 @@ class _HandleGeneratorArguments(VisitorTransform, SkipDeclarations):
                 cname = EncodedString(Naming.genexpr_arg_prefix + Symtab.punycodify_name(str(name_source)))
                 name_decl = Nodes.CNameDeclaratorNode(pos=pos, name=name)
                 type = node.type
-                while ((type.is_reference and not type.is_fake_reference) or
-                        type.is_cv_qualified):
-                    if type.is_cv_qualified:
-                        # strip away cv types - they shouldn't be applied to the
-                        # function argument or to the closure struct.
-                        type = type.cv_base_type
-                    else:
-                        # It isn't obvious whether the right thing to do would be to capture by reference or by
-                        # value (C++ itself doesn't know either for lambda functions and forces a choice).
-                        # However, capture by reference involves converting to FakeReference which would require
-                        # re-analysing AttributeNodes. Therefore I've picked capture-by-value out of convenience
-                        # TODO - could probably be optimized by making the arg a reference but the closure not
-                        # (see https://github.com/cython/cython/issues/2468)
-                        type = type.ref_base_type
+
+                # strip away cv types - they shouldn't be applied to the
+                # function argument or to the closure struct.
+                # It isn't obvious whether the right thing to do would be to capture by reference or by
+                # value (C++ itself doesn't know either for lambda functions and forces a choice).
+                # However, capture by reference involves converting to FakeReference which would require
+                # re-analysing AttributeNodes. Therefore I've picked capture-by-value out of convenience
+                # TODO - could probably be optimized by making the arg a reference but the closure not
+                # (see https://github.com/cython/cython/issues/2468)
+                type = PyrexTypes.remove_cv_ref(type, dont_remove_fakeref=True)
 
                 name_decl.type = type
                 new_arg = Nodes.CArgDeclNode(pos=pos, declarator=name_decl,

--- a/Cython/Compiler/PyrexTypes.py
+++ b/Cython/Compiler/PyrexTypes.py
@@ -5371,7 +5371,7 @@ def cap_length(s, max_prefix=63, max_len=1024):
     hash_prefix = hashlib.sha256(s.encode('ascii')).hexdigest()[:6]
     return '%s__%s__etc' % (hash_prefix, s[:max_len-17])
 
-def remove_cv_ref(tp, dont_remove_fakeref=True):
+def remove_cv_ref(tp, remove_fakeref=False):
     # named by analogy with c++ std::remove_cv_ref
     last_tp = None
     # The while-loop is probably unnecessary, but I'm not confident
@@ -5380,6 +5380,6 @@ def remove_cv_ref(tp, dont_remove_fakeref=True):
         last_tp = tp
         if tp.is_cv_qualified:
             tp = tp.cv_base_type
-        if tp.is_reference and not (dont_remove_fakeref and tp.is_fake_reference):
+        if tp.is_reference and (not tp.is_fake_reference or remove_fakeref):
             tp = tp.ref_base_type
     return tp

--- a/Cython/Compiler/PyrexTypes.py
+++ b/Cython/Compiler/PyrexTypes.py
@@ -5370,3 +5370,16 @@ def cap_length(s, max_prefix=63, max_len=1024):
         return s
     hash_prefix = hashlib.sha256(s.encode('ascii')).hexdigest()[:6]
     return '%s__%s__etc' % (hash_prefix, s[:max_len-17])
+
+def remove_cv_ref(tp, dont_remove_fakeref=True):
+    # named by analogy with c++ std::remove_cv_ref
+    last_tp = None
+    # The while-loop is probably unnecessary, but I'm not confident
+    # of the order or how careful we are prevent nesting.
+    while tp != last_tp:
+        last_tp = tp
+        if tp.is_cv_qualified:
+            tp = tp.cv_base_type
+        if tp.is_reference and not (dont_remove_fakeref and tp.is_fake_reference):
+            tp = tp.ref_base_type
+    return tp

--- a/Cython/Compiler/TypeInference.py
+++ b/Cython/Compiler/TypeInference.py
@@ -538,10 +538,7 @@ def find_spanning_type(type1, type2):
     return result_type
 
 def simply_type(result_type):
-    if result_type.is_reference:
-        result_type = result_type.ref_base_type
-    if result_type.is_cv_qualified:
-        result_type = result_type.cv_base_type
+    result_type = PyrexTypes.remove_cv_ref(result_type, dont_remove_fakeref=False)
     if result_type.is_array:
         result_type = PyrexTypes.c_ptr_type(result_type.base_type)
     return result_type

--- a/Cython/Compiler/TypeInference.py
+++ b/Cython/Compiler/TypeInference.py
@@ -538,7 +538,7 @@ def find_spanning_type(type1, type2):
     return result_type
 
 def simply_type(result_type):
-    result_type = PyrexTypes.remove_cv_ref(result_type, dont_remove_fakeref=False)
+    result_type = PyrexTypes.remove_cv_ref(result_type, remove_fakeref=True)
     if result_type.is_array:
         result_type = PyrexTypes.c_ptr_type(result_type.base_type)
     return result_type

--- a/tests/run/cpp_iterators.pyx
+++ b/tests/run/cpp_iterators.pyx
@@ -399,11 +399,11 @@ cdef extern from *:
 
 def test_iteration_from_const_member(int num):
     """
-    >>> test_iteration_from_const_member(5)
+    >>> tuple(test_iteration_from_const_member(5))
     (0, 1, 2, 3, 4)
     """
     num_holder = new ConstNumberHolder(num)
     try:
-        return tuple(i for i in range(num_holder.num))
+        return (i for i in range(num_holder.num))
     finally:
         del num_holder

--- a/tests/run/cpp_iterators.pyx
+++ b/tests/run/cpp_iterators.pyx
@@ -376,3 +376,34 @@ def test_non_built_in_reversed_function(py_v):
         vint.push_back(e)
     for e in reversed(vint):
         print(e)
+
+
+# Not strictly a C++ iterator test, but an issue with generator
+# expressions and iteration from c++ const attributes so in this
+# file as a related issue.  Mainly a test that it compiles.
+# GH-5558
+cdef extern from *:
+    """
+    struct ConstNumberHolder {
+        const int num;
+
+        explicit ConstNumberHolder(int num) :
+            num(num)
+        {}
+    };
+    """
+    cppclass ConstNumberHolder:
+        ConstNumberHolder(int num)
+
+        const int num
+
+def test_iteration_from_const_member(int num):
+    """
+    >>> test_iteration_from_const_member(5)
+    (0, 1, 2, 3, 4)
+    """
+    num_holder = new ConstNumberHolder(num)
+    try:
+        return tuple(i for i in range(num_holder.num))
+    finally:
+        del num_holder

